chore: add fdv2 store & update sources to use fdv2 protocol definitions

cwaldren-ld opened the pull request:
This adds the FDv2 dual-mode store, which supports serving data requests from an in-memory store or from a persistent store (with cache).

cwaldren-ld committed:
chore: add fdv2 store

cwaldren-ld committed:
remove old comment

cwaldren-ld committed:
introduce DataQuality enum

cwaldren-ld committed:
add SetPersist

cwaldren-ld committed:
use persist concept

cwaldren-ld committed:
make store implement FDv2 protocol

cwaldren-ld committed:
remove old data destination interface

cwaldren-ld committed:
revert chnages to datasource/streaming_data_source_events.go

cwaldren-ld commented on internal/datasource/streaming_data_source_events.go:
```diff
@@ -16,7 +16,7 @@ var (
 	deleteDataRequiredProperties = []string{"path", "version"} //nolint:gochecknoglobals
```
This is just a reversion of recent changes that made these public (for use in datasourcev2), but are no longer necessary.

cwaldren-ld committed:
use fdv2 types in streaming data source event handler

cwaldren-ld committed:
refactor concurrency tests

cwaldren-ld committed:
optimize comments on store.go

cwaldren-ld committed:
add dump method to memory store for atomically getting segments and features

cwaldren-ld commented on internal/datastore/in_memory_data_store_impl.go:
```diff
@@ -4,11 +4,10 @@ import (
 	"sync"
 
 	"github.com/launchdarkly/go-sdk-common/v3/ldlog"
-	"github.com/launchdarkly/go-server-sdk/v7/subsystems"
```
TODO: Update this component's unit tests with tests for SetBasis/ApplyDelta/Dump

cwaldren-ld committed:
inMemoryStore tests for SetBasis/ApplyDelta

cwaldren-ld committed:
add new memorystorev2 package

cwaldren-ld committed:
update tests

cwaldren-ld committed:
more unit tests

cwaldren-ld committed:
chore: introduce memorystorev2

cwaldren-ld committed:
benchmarks

cwaldren-ld committed:
lints

cwaldren-ld committed:
use defer pattern for locks

cwaldren-ld committed:
make element matcher consistent

cwaldren-ld committed:
doc tweaks

cwaldren-ld committed:
Merge branch 'cw/sdk-709/memory-store' into cw/sdk-557/fdv2-store

cwaldren-ld committed:
add better collection matcher

cwaldren-ld committed:
Merge branch 'v7' into cw/sdk-557/fdv2-store

cwaldren-ld committed:
make Selector a pointer type

cwaldren-ld commented on internal/datasourcev2/polling_data_source.go:
```diff
@@ -18,10 +18,47 @@ const (
 	pollingWillRetryMessage = "will retry at next scheduled poll interval"
 )
 
-// Requester allows PollingProcessor to delegate fetching data to another component.
```
A bit of new binding code so that the requestor can pass back FDv2 related data.

cwaldren-ld commented on internal/datasourcev2/polling_data_source.go:
```diff
@@ -142,16 +179,23 @@ func (pp *PollingProcessor) Start(closeWhenReady chan<- struct{}) {
+	if response.Cached() {
+		return nil
 	}
+
```
Use the new FDv2 store APIs!

cwaldren-ld committed:
cleanup persistent store insertion logic

cwaldren-ld committed:
lints

cwaldren-ld committed:
more lints

cwaldren-ld committed:
Merge branch 'v7' into cw/sdk-557/fdv2-store

cwaldren-ld committed:
attempt to suppress TODO lint failures

cwaldren-ld commented on internal/datasourcev2/polling_http_request.go:
```diff
@@ -68,42 +71,110 @@ func (r *pollingRequester) BaseURI() string {
+				}
+
+				intent = serverIntent.Payloads[0].Code
+				if intent == "none" {
```
We have a constant for this (or should)

cwaldren-ld commented on internal/datasourcev2/polling_http_request.go:
```diff
@@ -68,42 +71,110 @@ func (r *pollingRequester) BaseURI() string {
+				for obj := r.Object().WithRequiredProperties([]string{versionField, kindField, "key", "object"}); obj.Next(); {
+					switch string(obj.Name()) {
+					case versionField:
+						// version = r.Int()
```
We can take the version now

cwaldren-ld commented on internal/datasourcev2/polling_http_request.go:
```diff
@@ -68,42 +71,110 @@ func (r *pollingRequester) BaseURI() string {
+					case kindField:
+						kind = strings.TrimRight(r.String(), "s")
+						dataKind = dataKindFromKind(kind)
+					case "key":
```
We have consts for this now

cwaldren-ld commented on internal/datasourcev2/polling_http_request.go:
```diff
@@ -68,42 +71,110 @@ func (r *pollingRequester) BaseURI() string {
+						// version = r.Int()
+					case kindField:
+						kind = strings.TrimRight(r.String(), "s")
+						dataKind = dataKindFromKind(kind)
```
This needs to use the new kind deserializer

cwaldren-ld commented on internal/datasourcev2/streaming_data_source.go:
```diff
@@ -249,36 +242,29 @@ func (sp *StreamProcessor) consumeStream(stream *es.Stream, closeWhenReady chan<
+				switch currentChangeSet.intent.Payloads[0].Code {
+				case fdv2proto.IntentTransferFull:
+					{
+						_ = sp.dataDestination.SetBasis(updates, fdv2proto.NoSelector(), true)
```
Shouldn't return errors here. Should handle in data system, then use DataStoreStatus or other mechanism for testing / reporting

cwaldren-ld commented on internal/datasystem/store.go:
```diff
@@ -0,0 +1,258 @@
+
+	// Source of truth for flag evaluations (once initialized). Before initialization,
+	// the persistentStore may be used if configured.
+	memoryStore *memorystorev2.Store
```
import alias `memory`?

cwaldren-ld commented on internal/datasystem/store.go:
```diff
@@ -0,0 +1,258 @@
+	return s.init(collections, selector, persist)
+}
+
+func (s *Store) init(allData []ldstoretypes.Collection, selector *fdv2proto.Selector, persist bool) error {
```
Move this into SetBasis. We don't need this.

cwaldren-ld commented on internal/datasystem/store.go:
```diff
@@ -0,0 +1,258 @@
+	return s.active
+}
+
+//nolint:revive // Implementation for ReadOnlyStore.
```
Can we return a pointer to the right store once for each evaluation?

cwaldren-ld commented:
@keelerm84 , realized why I was confused - the data system config PR has some of the other improvements in the streaming/polling sources (like replacing strings with constants). I tried to only touch the sources enough to get them to interact with the new store. 

I'll have another PR that focuses on cleaning up the sources.

cwaldren-ld committed:
move store.init() into store.SetBasis

cwaldren-ld committed:
no need to inspect error from ApplyDelta/SetBasis

cwaldren-ld committed:
remove old TODO

keelerm84 commented on internal/datasourcev2/polling_http_request.go:
```diff
@@ -68,42 +71,110 @@ func (r *pollingRequester) BaseURI() string {
-	for _, event := range payload.Events {
-		esEvents = append(esEvents, event)
+	parseItem := func(r jreader.Reader, kind datakinds.DataKindInternal) (ldstoretypes.ItemDescriptor, error) {
+		item, err := kind.DeserializeFromJSONReader(&r)
+		return item, err
```
I believe this is equivalent:

```suggestion
		return kind.DeserializeFromJSONReader(&r)
```

Which makes `parseItem` a single line method. At that point, is there value in having it as a a separate function?

keelerm84 commented on internal/datasourcev2/polling_http_request.go:
```diff
@@ -68,42 +71,110 @@ func (r *pollingRequester) BaseURI() string {
-	if !ok {
-		return nil, false, malformedJSONError{errors.New("payload is not a PutData")}
-	}
+	var intent fdv2proto.IntentCode
```
Can we consider changing this name to `intentCode` or something? If I see `intent`, I'm going to think the `serverIntent` message more than the code.

keelerm84 commented on internal/datasourcev2/polling_response.go:
```diff
@@ -0,0 +1,48 @@
+type PollingResponse struct {
+	events   []fdv2proto.Event
+	cached   bool
+	intent   fdv2proto.IntentCode
```
I would think we would want to rename this as well. But I could be swayed I suppose. 😄

keelerm84 commented on internal/datasourcev2/streaming_data_source.go:
```diff
@@ -449,25 +435,17 @@ func (sp *StreamProcessor) GetFilterKey() string {
 	for _, event := range events {
-		switch event.Event() {
-		case putEventName:
+		switch fdv2proto.EventName(event.Event()) {
```
I can't comment on the actual line, but `parseItem` here suffers from the same issue I pointed out above.

cwaldren-ld replied:
It turns out this whole bit of logic (the event parsing) doesn't work properly due to the streaming JSON parser. I'm going to have to refactor this to more closely match how the existing datasource (V1) parsers work - which is two passes. 

The reason is because we might visit the "object" field _before_ we've visited the kind field.

keelerm84 commented on internal/datasystem/store.go:
```diff
@@ -0,0 +1,252 @@
+	"github.com/launchdarkly/go-server-sdk/v7/subsystems/ldstoretypes"
+)
+
+// Store is a dual-mode persistent/in-memory store that data queries from the evaluation
+// algorithm.
```
`that data queries from` doesn't make sense to me. Can we re-word that?

cwaldren-ld replied:
Ah, was missing a word. 👍

keelerm84 commented on internal/datasystem/store.go:
```diff
@@ -0,0 +1,252 @@
+// store.
+//
+// One motivation behind using persistent stores is to offer a way to immediately start evaluating
+// flags before a connection is made to LD (or even in the moment before an initializer has run.)
```
Sorry for the nitpick.

```suggestion
// flags before a connection is made to LD (or even in the moment before an initializer has run).
```

keelerm84 commented on internal/fdv2proto/event_to_storable_item.go:
```diff
@@ -0,0 +1,59 @@
+func ToStorableItems(events []Event) []ldstoretypes.Collection {
+	flagCollection := ldstoretypes.Collection{
+		Kind:  datakinds.Features,
+		Items: make([]ldstoretypes.KeyedItemDescriptor, 0),
```
Do you think it's worth reserving some capacity here to prevent the additional allocations?

cwaldren-ld replied:
We don't know the size up front, although we could calculate it.  I think I'd rather let Go's allocation strategy do its thing and if it's a problem we can fix it.

cwaldren-ld committed:
refactor: more fdv2 protocol types + usage in data sources (#198)

This fully converts the datasources to use `fdv2` protocol domain
objects.

cwaldren-ld committed:
Merge branch 'v7' into cw/sdk-557/fdv2-store

cwaldren-ld committed:
refactor: insert items into fdv2 store in dependency order (#197)

Use the new `toposort` package to insert items int persistent store in
dependency order.

cwaldren-ld committed:
code review: fixing comments

cwaldren-ld committed:
lints

cwaldren-ld committed:
remove circular dependency in toposort
